src/backend/distributed/executor/intermediate_result_encoder.c

```diff
@@ -0,0 +1,324 @@
+	CopyOutState copyOutState;
+	FmgrInfo *columnOutputFunctions;
+};
+
+/* forward declaration of local functions */
+static void ReadCopyFileIntoTupleStore(char *fileName, char *copyFormat,
+									   TupleDesc tupleDescriptor,
+									   Tuplestorestate *tupstore);
+static Relation StubRelation(TupleDesc tupleDescriptor);
+
+
+/*
+ * IntermediateResultEncoderCreate returns an encoder which can encode tuples of
+ * given tupleDesc form using the given format. Encoder possibly can add header
+ * data at this stage.
+ */
+IntermediateResultEncoder *
+IntermediateResultEncoderCreate(TupleDesc tupleDesc,
+								IntermediateResultFormat format,
+								MemoryContext tupleContext)
+{
+	IntermediateResultEncoder *encoder = palloc0(sizeof(IntermediateResultEncoder));
+	encoder->format = format;
+	encoder->tupleDescriptor = CreateTupleDescCopy(tupleDesc);
+	encoder->outputBuffer = makeStringInfo();
+	encoder->flushBuffer = makeStringInfo();
+
+	if (format == TEXT_COPY_FORMAT || format == BINARY_COPY_FORMAT)
+	{
+		int fileEncoding = pg_get_client_encoding();
+		int databaseEncoding = GetDatabaseEncoding();
+		int databaseEncodingMaxLength = pg_database_encoding_max_length();
+
+		char *nullPrint = "\\N";
+		int nullPrintLen = strlen(nullPrint);
+		char *nullPrintClient = pg_server_to_any(nullPrint, nullPrintLen, fileEncoding);
+
+		CopyOutState copyOutState = palloc0(sizeof(CopyOutStateData));
+		copyOutState->delim = "\t";
+		copyOutState->null_print = nullPrint;
+		copyOutState->null_print_client = nullPrintClient;
+		copyOutState->binary = (format == BINARY_COPY_FORMAT);
+		copyOutState->fe_msgbuf = encoder->outputBuffer;
+		copyOutState->rowcontext = tupleContext;
+
+		if (PG_ENCODING_IS_CLIENT_ONLY(fileEncoding))
+		{
+			ereport(ERROR, (errmsg("cannot repartition into encoding caller "
+								   "cannot receive")));
+		}
+
+		/* set up transcoding information and default text output characters */
+		copyOutState->need_transcoding = (fileEncoding != databaseEncoding) ||
+										 (databaseEncodingMaxLength > 1);
+
+		encoder->copyOutState = copyOutState;
+		encoder->columnOutputFunctions =
+			ColumnOutputFunctions(tupleDesc, copyOutState->binary);
+
+		if (copyOutState->binary)
+		{
+			AppendCopyBinaryHeaders(copyOutState);
+		}
+	}
+
+	return encoder;
+}
+
+
+/*
+ * IntermediateResultEncoderReceive encodes the next row with the given encoder.
+ */
+StringInfo
+IntermediateResultEncoderReceive(IntermediateResultEncoder *encoder,
+								 Datum *values, bool *nulls)
+{
+	if (encoder->format == TEXT_COPY_FORMAT ||
+		encoder->format == BINARY_COPY_FORMAT)
+	{
+		AppendCopyRowData(values, nulls, encoder->tupleDescriptor,
+						  encoder->copyOutState, encoder->columnOutputFunctions, NULL);
+	}
+
+	if (encoder->outputBuffer->len > ENCODER_BUFFER_SIZE)
+	{
+		encoder->flushBuffer->data = encoder->outputBuffer->data;
+		encoder->flushBuffer->len = encoder->outputBuffer->len;
+
+		encoder->outputBuffer->len = 0;
```

Comment:
This seems a bit gross. Aliasing of `StringInfo` data means length is going to be screwed up when either has operations done on it

Tough call, but there's essentially a "flushing" state that we enter & assume is done by time we start writing to outputBuffer again. Ideally we'd have a `bool flushing` which gets set, & then `Assert(!flushing)` everywhere with a `IntermediateResultEncoderFlushed` call to clear flushing. We essentially want to setup an API where we give out ownership of the StringInfo to the caller & the caller returns the StringInfo to us after it's flushed it. There's no need for two StringInfo fields in the encoder

Reply:
Could we replace StringInfo by something else that's similar, but doesn't have all these useful functions that could wreak havoc in this case? It's always been a little weird to use StringInfo for binary data as well.